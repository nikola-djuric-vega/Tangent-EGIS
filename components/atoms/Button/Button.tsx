import clsx from 'clsx'
import React, { cloneElement, useEffect, useState } from 'react'

import tailwindConfig, { theme } from '../../../tailwind.config'

interface Props {
  type: 'primary' | 'secondary' | 'tertiary'
  buttonText?: string
  backgroundColour?: string
  onClick(): void
  smallText?: boolean
  icon?: any
  hoverColour?: string
  textColour?: string
  textHoverColour?: string
  onlyBorder?: boolean
  onlyBorderWhite?: boolean
  active?: boolean
}

export default function Button({
  type,
  buttonText,
  backgroundColour,
  onClick,
  smallText,
  icon,
  hoverColour,
  textColour,
  textHoverColour,
  onlyBorder,
  onlyBorderWhite,
  active,
}: Props) {
  const [isMouseOver, setIsMouseOver] = useState<boolean>(false)
  const [arrowColor, setArrowColor] = useState<string | any>('')

  useEffect(() => {
    if (textHoverColour && textColour) {
      let stripTextHover = textHoverColour.split('text-').pop()
      let stripTextColour = textColour.split('text-').pop()

      if (isMouseOver) {
        setArrowColor(tailwindConfig.theme.colors[stripTextHover!])
      } else {
        setArrowColor(tailwindConfig.theme.colors[stripTextColour!])
      }
    }
  }, [isMouseOver, textHoverColour, textColour])

  return (
    <button
      data-testid="button-item"
      onMouseOver={() => setIsMouseOver(true)}
      onMouseLeave={() => setIsMouseOver(false)}
      onClick={() => onClick()}
      className={clsx(
        'rounded-full inline-flex flex-nowrap whitespace-nowrap items-center justify-center ease-in-out transition duration-500 group cursor-pointer font-extrabold',
        active
          ? 'bg-midnight-blue'
          : isMouseOver && textHoverColour
          ? hoverColour
          : backgroundColour,

        {
          'border-2 border-midnight-blue hover:bg-midnight-blue font-extrabold':
            onlyBorder,
          'bg-white': onlyBorder && !active,
          'border-2 border-white': onlyBorderWhite,
          'py-3.5 px-6': type === 'primary',
          'py-2 px-5': type === 'secondary',
          'border-none bg-opacity-0 p-0': type === 'tertiary',
        }
      )}
    >
      <span
        className={clsx(
          active ? 'text-white' : textColour && textColour,
          textHoverColour && textHoverColour,
          {
            'bg-opacity-0 hover:underline': type === 'tertiary',
            'text-xs': smallText,
          }
        )}
      >
        {buttonText}
      </span>
      {icon && (
        <div className="ml-3">
          {cloneElement(icon, {
            arrowColor: active ? theme.colors['white'] : arrowColor,
          })}
        </div>
      )}
    </button>
  )
}
