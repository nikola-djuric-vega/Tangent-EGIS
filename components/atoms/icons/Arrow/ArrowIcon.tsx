import * as React from 'react'

import clsx from 'clsx'

import { theme } from '../../../../tailwind.config'
import { SvgProps } from '../../../../types'

interface Props extends SvgProps {
  backgroundColor?: 'super-lime' | 'teal-blue'
  arrowColor?: string
  rightArrow?: boolean
  opacity?: boolean
  angle?: boolean
}

export default function ArrowIcon({
  width = 46,
  height = 46,
  backgroundColor,
  arrowColor,
  rightArrow = false,
  opacity,
  angle,
  ...props
}: Props) {
  return (
    <div
      className={clsx(
        'rounded-full flex items-center justify-center bg-opacity-0 transition duration-500 ease-out hover:bg-transparent border-2 box-content border-transparent',
        {
          'bg-super-lime bg-opacity-100 p-4 hover:border-super-lime':
            backgroundColor === 'super-lime',
          'bg-teal-blue bg-opacity-20 p-4 hover:border-teal-blue hover:border-opacity-40':
            backgroundColor === 'teal-blue',
        }
      )}
      style={{ width: width, height: height }}
    >
      <svg
        className={clsx(
          'transform flex justify-center items-center w-full h-full',
          rightArrow ? 'rotate-180' : angle ? 'rotate-135' : 'rotate-0'
        )}
        viewBox="0 0 14 10"
        fill="none"
        xmlns="http://www.w3.org/2000/svg"
        {...props}
      >
        <path
          d="M4.58 3.695h8.92v2.596l-8.92.014L7.29 10H4.166L.5 4.993 4.166 0 7.29.013 4.58 3.695z"
          fill={arrowColor ? arrowColor : theme.colors['midnight-blue']}
        />
      </svg>
    </div>
  )
}
