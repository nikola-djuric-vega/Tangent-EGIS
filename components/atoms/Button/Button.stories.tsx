import React from 'react'
import { ComponentStory, ComponentMeta } from '@storybook/react'

import Button from './Button'
import ArrowIcon from '../icons/Arrow/ArrowIcon'

export default {
  title: 'Atoms/Button',
  component: Button,
} as ComponentMeta<typeof Button>

const Template: ComponentStory<typeof Button> = (args) => <Button {...args} />

export const Primary = Template.bind({})
Primary.args = {
  type: 'primary',
  buttonText: 'Call to action label',
  backgroundColour: 'bg-super-lime',
  hoverColour: 'bg-midnight-blue',
  textColour: 'text-midnight-blue',
  textHoverColour: 'text-super-lime',
  icon: <ArrowIcon rightArrow width={17} height={20} />,
}

export const Secondary = Template.bind({})
Secondary.args = {
  type: 'secondary',
  buttonText: 'Call to action label',
  textColour: 'text-midnight-blue',
  textHoverColour: 'text-white',
  onlyBorder: true,
  icon: <ArrowIcon rightArrow width={17} height={20} />,
}

export const Tertiary = Template.bind({})
Tertiary.args = {
  type: 'tertiary',
  buttonText: 'Call to action label',
}

export const TertiarySmallText = Template.bind({})
TertiarySmallText.args = {
  type: 'tertiary',
  buttonText: 'Call to action label',
  smallText: true,
}
