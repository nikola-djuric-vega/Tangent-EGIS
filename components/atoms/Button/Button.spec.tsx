// Button.test.tsx (Note that tests are colocated with components)
import React from 'react'
import { render, fireEvent, screen, waitFor } from '@testing-library/react'
import Button from './Button'
import ArrowIcon from '../icons/Arrow/ArrowIcon'

describe('ButtonComponent', () => {
  test('it should render a primary button', async () => {
    const mockClick = jest.fn()
    const { getByText } = render(
      <Button
        onClick={mockClick}
        type="primary"
        buttonText="primary"
        backgroundColour="bg-super-lime"
        hoverColour="bg-midnight-blue"
        textColour="text-midnight-blue"
        textHoverColour="group-hover:text-super-lime"
      />
    )
    const TestButton = getByText('primary')
    expect(TestButton).toBeInTheDocument()
    expect(TestButton).toMatchSnapshot()
  })

  test('it should render a primary button with a icon', async () => {
    const mockClick = jest.fn()
    const { getByText } = render(
      <Button
        type="primary"
        buttonText="primary"
        textColour="text-midnight-blue"
        textHoverColour="group-hover:text-white"
        hoverColour="bg-midnight-blue"
        onlyBorder
        onClick={mockClick}
        icon={<ArrowIcon rightArrow width={17} height={20} />}
      />
    )
    const TestButton = getByText('primary')
    expect(TestButton).toBeInTheDocument()
    expect(TestButton).toMatchSnapshot()
  })

  test('it should render a secondary button', async () => {
    const mockClick = jest.fn()
    const { getByText } = render(
      <Button
        type="primary"
        buttonText="secondary"
        backgroundColour="bg-super-lime"
        hoverColour="bg-midnight-blue"
        textColour="text-midnight-blue"
        textHoverColour="group-hover:text-super-lime"
        onClick={mockClick}
        icon={<ArrowIcon rightArrow width={17} height={20} />}
      />
    )
    const TestButton = getByText('secondary')
    expect(TestButton).toBeInTheDocument()
    expect(TestButton).toMatchSnapshot()
  })

  test('it should render a tertiary button', async () => {
    const mockClick = jest.fn()
    const { getByText } = render(
      <Button type="tertiary" buttonText="tertiary" onClick={mockClick} />
    )
    const TestButton = getByText('tertiary')
    expect(TestButton).toBeInTheDocument()
    expect(TestButton).toMatchSnapshot()
  })

  test('it should perform the passed onClick function', async () => {
    const mockClick = jest.fn()
    const { getByText } = render(
      <Button
        onClick={mockClick}
        type="primary"
        buttonText="onClick"
        backgroundColour="bg-super-lime"
        hoverColour="bg-midnight-blue"
        textColour="text-midnight-blue"
        textHoverColour="group-hover:text-super-lime"
        icon={<ArrowIcon rightArrow width={17} height={20} />}
      />
    )
    fireEvent.click(getByText('onClick'))
    expect(mockClick.mock.calls.length).toBe(1)
  })

  test('it should perform mouse over and change background colour to hover colour', async () => {
    const mockClick = jest.fn()
    const { container } = render(
      <Button
        onClick={mockClick}
        type="primary"
        buttonText="mousehover"
        backgroundColour="bg-super-lime"
        hoverColour="hover:bg-midnight-blue"
        textColour="text-midnight-blue"
        textHoverColour="group-hover:text-super-lime"
        icon={<ArrowIcon rightArrow width={17} height={20} />}
      />
    )
    fireEvent.mouseOver(screen.getByText('mousehover'))
    await waitFor(() => screen.getByTestId('button-item'))
    expect(container.firstChild).toHaveClass('hover:bg-midnight-blue')
  })

  test('it should perform mouse leave and change colour back to background colour', async () => {
    const mockClick = jest.fn()
    const { container } = render(
      <Button
        onClick={mockClick}
        type="primary"
        buttonText="mousehover"
        backgroundColour="bg-super-lime"
        hoverColour="bg-midnight-blue"
        textColour="text-midnight-blue"
        textHoverColour="group-hover:text-super-lime"
        icon={<ArrowIcon rightArrow width={17} height={20} />}
      />
    )
    fireEvent.mouseLeave(screen.getByText('mousehover'))
    await waitFor(() => screen.getByTestId('button-item'))
    expect(container.firstChild).toHaveClass('bg-super-lime')
  })
})
