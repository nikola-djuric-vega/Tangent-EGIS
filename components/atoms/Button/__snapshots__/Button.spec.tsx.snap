// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ButtonComponent it should render a primary button 1`] = `
<span
  class="text-midnight-blue group-hover:text-super-lime"
>
  primary
</span>
`;

exports[`ButtonComponent it should render a primary button with a icon 1`] = `
<span
  class="text-midnight-blue group-hover:text-white"
>
  primary
</span>
`;

exports[`ButtonComponent it should render a secondary button 1`] = `
<span
  class="text-midnight-blue group-hover:text-super-lime"
>
  secondary
</span>
`;

exports[`ButtonComponent it should render a tertiary button 1`] = `
<span
  class="bg-opacity-0 hover:underline"
>
  tertiary
</span>
`;
